### PR TITLE
adding support for streaming results via data channels

### PR DESCRIPTION
**Pull request by isaacrob**
We want to enable real-time communication with the DirectAI backend. To do that, we can send tracking results to the client directly via WebRTC data channels, avoiding the use of webhooks.

### CODE REVIEW

File: README.md
Comment (ben-directai) [inline README.md]: to clarify: the data channel requires a separate livekit connection with its own set of credentials. this connection _does not_ generate a third room attendee

File: README.md
Comment (ben-directai) [inline README.md]: i'm assuming this sentence was meant to include more words?

File: stream_script.py
Comment (ben-directai) [inline stream_script.py]: i'm assuming we don't want this to be public

File: result_pipe.py
Comment (ben-directai) [inline result_pipe.py:16]: any particular reason these are denoted as "topics"? I'm assuming Livekit labels them this way?

File: result_pipe.py
Comment (ben-directai) [inline result_pipe.py:47]: to clarify: we create a new uuid for _every_ message, not just for each stream?

File: result_pipe.py
Comment (ben-directai) [reply result_pipe.py:16]: now seeing `publish_data()` expects a topic param

File: result_pipe.py
Comment (ben-directai) [inline result_pipe.py:106]: nit: not sure we need this

File: stream_script.py
Comment (ben-directai) [inline stream_script.py]: is `LIVEKIT_TOKEN_FOR_RESULTS` None or "" by default?

File: stream_script.py
Comment (ben-directai) [inline stream_script.py:186]: nit: I'd expect KeyboardInterrupt to be in the following sequence:
```
try:

except asyncio.CancelledError: 

except KeyboardInterrupt:
```

**Review by ben-directai — CHANGES_REQUESTED** [review]
mostly questions, wanted to update a couple lines before merging. i'm excited to see a demo on the server side PR!

File: README.md
Comment (isaacrob) [reply README.md]: this connection does create a third room attendee, which just listens for events. it's akin to having someone join google meets with no video or audio, just looking at the chat

File: stream_script.py
Comment (isaacrob) [reply stream_script.py]: it's a local network URL, not accessible outside our VPN, so should be fine .. although perhaps not best form, maybe this should be taken from the .env file? thought it might be confusing though to say that the URL is an env variable ..

File: result_pipe.py
Comment (isaacrob) [reply result_pipe.py:16]: yep livekit nomenclature

File: README.md
Comment (ben-directai) [reply README.md]: I was under the impression that there was a way for a token to be issued s.t. the "process" monitoring that room wouldn't be visible to Livekit viewers. Having an extra attendee with a dark screen isn't great from the user experience perspective - I'm willing to address in a different PR but wanted to point it out

File: result_pipe.py
Comment (isaacrob) [reply result_pipe.py:47]: I believe that is what's happening yes

File: result_pipe.py
Comment (isaacrob) [reply result_pipe.py:106]: going by the livekit boilerplate code for a chat manager (which this borrows from), it's actually non obvious where the application logic should live .. I think this is helpful

File: stream_script.py
Comment (isaacrob) [reply stream_script.py]: it's None if LIVEKIT_TOKEN_FOR_RESULTS is not in the .env file, but '' if it's in the file but just blank or just whitespace .. I guess I should have it default to '' and catch that then?

File: stream_script.py
Comment (isaacrob) [reply stream_script.py:186]: it doesn't work there because of internals of asyncio, has to be caught outside the loop

File: README.md
Comment (isaacrob) [reply README.md]: yeah .. my bad, updated

File: stream_script.py
Comment (ben-directai) [reply stream_script.py]: i'm fine with hardcoding the base url or setting the url in the env file, just don't want to merge the local IP to `main`

File: result_pipe.py
Comment (ben-directai) [reply result_pipe.py:106]: kk

File: stream_script.py
Comment (ben-directai) [reply stream_script.py]: maybe both - I could imagine user failure in either direction

File: stream_script.py
Comment (ben-directai) [reply stream_script.py:186]: sg

File: README.md
Comment (isaacrob) [reply README.md]: yeah, that's controlled on the token generation side, not our side .. there's a checkbox they can check that will make the person invisible when they issue the token

File: stream_script.py
Comment (isaacrob) [reply stream_script.py]: oh I forgot that this was in the version I was trying to merge .. my bad .. I'll leave it hardcoded pointing to free for now

File: stream_script.py
Comment (isaacrob) [reply stream_script.py]: it's None because I'm saying in the .get to return None in the case where it's missing .. if I return '' in the case where it's missing I can catch both modes in one go I think

File: README.md
Comment (ben-directai) [inline README.md]: nit: second sentence is run-on

suggested rephrasing: If you include a token in `LIVEKIT_TOKEN_FOR_RESULTS`, you'll see real-time object tracking results and end-to-end latency measurements returned to the client via WebRTC data channels.

File: stream_script.py
Comment (ben-directai) [reply stream_script.py]: yeah that sounds good!

File: README.md
Comment (ben-directai) [reply README.md]: i'd like to make a note in the docs for this. again, can be done later (#3)

File: README.md
Comment (isaacrob) [reply README.md]: don't think it's a run-on, just think it's a bad sentence haha .. it's important to note that it is NOT end to end latency in the traditional sense because we don't have a measure of at what time the image was transmitted from the user's machine. it's instead a measure of the time from when DirectAI received the image from livekit to when the client received the result. 

how about: "If you included a token in `LIVEKIT_TOKEN_FOR_RESULTS`, you'll see real-time object tracking results returned to the client via WebRTC data channels. In addition, you'll see a measure of latency from when DirectAI received the frame from LiveKit to when the client received the tracking results from the data channel."

File: stream_script.py
Comment (isaacrob) [reply stream_script.py]: worth noting that this URL won't work until we've merged the backend changes to prod .. perhaps we should just wait to merge this into main until the listed URL works?

Comment (ben-directai) [reply stream_script.py]: yeah I'm fine waiting to merge until the URL works

File: README.md
Comment (ben-directai) [reply README.md]: makes sense re: distinction with e2e latency

a couple notes on the revised version:
- I'd prefer "If you include" as opposed to "If you included" as it's a present/future instruction
- Where is the measure of latency being sent to? I assume the webrtc data channel but it's not obvious from my reading

Comment (isaacrob) [reply README.md]: it's not being sent anywhere, it's being measured and printed in this code

**Review by ben-directai — APPROVED** [review]
lgtm pending live URL communicating properly

**Pull request merged**
